README
NAME
    Device::BlinkyTape - Control a BlinkyTape led strip

SYNOPSIS
        use Device::BlinkyTape::WS2811; # BlinkyTape uses WS2811
        my $bb = Device::BlinkyTape::WS2811->new(dev => '/dev/tty.usbmodem');
        # Set all led pixels on full white
        $bb->all_on();
        # Set all led pixels off
        $bb->all_off();

        # Send a white pixel (RGB 255/255/255).
        # Pixels are sent one by one from left to right.
        $bb->send_pixel(255,255,255);
        $bb->send_pixel(5,5,5);
        # Show all sent pixels and reset send_pixel to the first pixel.
        $bb->show();

  dev
    The device where your usb ledstrip is at. Defaults to /dev/tty.usbmodem.

  port
    Instead of giving the device you can create the instance of this module
    by directly giving it a Device:SerialPort object. By default the
    Device::SerialPort object is created from the device given in the 'dev'
    parameter.

  gamma
    Specify the gamma correction. Defaults to [1,1,1]

  led_count
    Specify the led count, counting from 1. The default is 60.

  simulate
    Specify if the module should simulate a BlinkyTape onscreen instead of
    using one in the usb port. Defaults to 0 (false). If this is true then
    the port and dev parameters have no effect.

  all_on
    Turns all leds on.

  all_off
    Turns all leds off.

  send_pixel(r,g,b)
    Send the RGB value for the next pixel. Values 0-254 are sent as is,
    value 255 is converted to 254.

  show
    Shows the sent pixels and resets the send_pixel to the first led pixel
    of the strip. This is done by sending a single 255 value byte to the led
    strip.

ABOUT BLINKYTAPE
    Blinkytape is a controllable led strip available at
    http://blinkiverse.com/blinkytape/

USING THE MODULE ON OS X WITHOUT OWNING A BLINKYTAPE
    This module comes with a BlinkyTape simulator. Install X11 server to use
    the simulator: http://xquartz.macosforge.org/landing/

BUGS
    The device is not yet available so the module has been implemented by
    inspecting partly undocumented and unfinished code in other languages.
    Feel free to file any bug reports in Github, patches welcome.

REFERENCE READING
    Communicating with the Arduino in Perl
    http://playground.arduino.cc/interfacing/PERL

    Perl communication to Arduino over serial USB
    http://www.windmeadow.com/node/38

AUTHOR
    Oskari Okko Ojala <okko@cpan.org>

    Based on exampls/Blinkyboard.py in Blinkiverse's BlinkyTape repository
    at https://github.com/blinkiverse/BlinkyTape/ by Max Henstell
    (mhenstell) and Marty McGuire (martymcguire).

COPYRIGHT AND LICENSE
    Copyright (C) Oskari Okko Ojala 2013

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.8.8 or, at
    your option, any later version of Perl you may have available.



Changes
0.005     2015-03-07 00:30:00 Europe/Helsinki
    * Initial support for Windows. May or may not work.

0.004     2013-12-14 00:06:00 Europe/Helsinki
    * Flush input to prevent slowing down.
    * Device as the first param to for example/christmas-blinking.pl

0.003     2013-12-12 22:47:00 Europe/Helsinki

    * Send the show command in a triplet so that BlinkyTape doesn't wait for the next two bytes.
    * Added a new example and refined the earlier example.

0.002     2013-12-12 00:51:00 Europe/Helsinki

    * Fixes a freeze when bytes were sent too fast.

0.001     2013-07-04 00:30:00 Europe/Helsinki

    * Initial release.


example/christmas-blinking.pl
#!/usr/bin/env perl
use strict;
use lib '../lib';
use Device::BlinkyTape::WS2811; # BlinkyTape uses WS2811
use Time::HiRes qw/ usleep /;

my $devpeek = $ARGV[0];

if (!$devpeek) {
    print "\n";
    print "Usage: $0 [device ..] [amount_of_blinking] [blinking_speed_min] [blinking_speed_max]\n";
    print "   [device] is the file or files to BlinkyTape, for example /dev/tty.usbmodem1411\n";
    print "\n";
    print "   Default is            $0 [device] 30  1 30\n";
    print "   For faster blinking:  $0 [device] 30 30 30\n";
    print "   For lots of blinking  $0 [device]  1  1 30\n";
    print "\n";
    exit 255;
}

my @dev;
while ($ARGV[0] =~ /^\/dev/) {
    my $sdev = shift @ARGV;
    push @dev, $sdev;
}
use Data::Dumper;
print 'Devices: '.Dumper(\@dev);

my $amount_of_blinking = shift @ARGV || 30;
my $brightness_step_min = shift @ARGV || 1;
my $brightness_step_max = shift @ARGV || 30;

my @bb;
foreach my $sdev (@dev) {
    my $sbb = Device::BlinkyTape::WS2811->new(dev => $sdev);
    push @bb, $sbb;
}

my $device_count = scalar(@dev);
my $maxled = (60 * $device_count - 1);

my @led;
for (my $a=0; $a<=$maxled; $a++) {
    $led[$a] = 0;
}
my @ledb;
my @ledr;

for (my $b=0; $b<=100000; $b++) {
    if ($b % $amount_of_blinking == 0) {
        my $pos = int(rand($maxled)+1);
        if ($led[$pos]==0) { # if not already blinking
            my $brightness_step = int(rand($brightness_step_max-$brightness_step_min))+$brightness_step_min;
            $ledb[$pos] = $brightness_step;
            $ledr[$pos] = int(rand(254));
            $led[$pos] = $ledb[$pos]; # initiate blinking
        }
    }

    for (my $a=0; $a<=$maxled; $a++) {
        if ($led[$a]>0) {
            $led[$a] = $led[$a]+$ledb[$a];
        }

        $led[$a] = 0 if ($led[$a]>512);

        my $sendval = $led[$a];
        if ($led[$a]>255) {
            $sendval =512 - $sendval;
        }

        my $o = $sendval-$ledr[$a]; # redness
        $o=0 if ($o<0);
        
        $bb[$a % $device_count]->send_pixel($sendval, $o, $o);
    }
    foreach my $sbb (@bb) {
        $sbb->show(); # shows the sent pixel row.
    }
    usleep(400);
}



lib/Device/BlinkyTape/WS2811.pm
package Device::BlinkyTape::WS2811;
use strict;
BEGIN {
    # AUTHORITY
    # VERSION
}
use Moose;
use utf8;
use Time::HiRes qw(usleep);

extends 'Device::BlinkyTape';

=for Pod::Coverage send_pixel show gamma

=head1 NAME

Device::BlinkyTape:WS2811 - Control a WS2811-based BlinkyTape

=head1 SYNOPSIS

    use Device::BlinkyTape::WS2811;
    my $bb = Device::BlinkyTape::WS2811->new(dev => '/dev/tty.usbmodem');

See Device::BlinkyTape for documentation.

=cut

sub send_pixel {
    my $self = shift;
    my ($r, $g, $b) = (shift, shift, shift);
    $r = 254 if ($r == 255); # The 255 means end of led line and applies the colors. Drop that value by one. Blinkyboard.py does this.
    $g = 254 if ($g == 255);
    $b = 254 if ($b == 255);
    $self->port->write(chr($r));
    usleep($self->sleeptime);
    $self->port->write(chr($g));
    usleep($self->sleeptime);
    $self->port->write(chr($b));
    usleep($self->sleeptime);
    my $string_in = $self->port->input; # flush input to prevent slowing down
}



sub show {
    my $self = shift;
    $self->port->write(chr(0));
    usleep($self->sleeptime);
    $self->port->write(chr(0));
    usleep($self->sleeptime);
    $self->port->write(chr(255));
    usleep($self->sleeptime);
    my $string_in = $self->port->input; # flush input to prevent slowing down
}

sub gamma {
    my $self = shift;
    my $input = shift;
    my $tweak = shift;
    return $input if not $self->gamma;
    return int(($input/256 ** $tweak) * 256);
}

1;

=head1 AUTHOR

Oskari Okko Ojala E<lt>okko@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) Oskari Okko Ojala 2013

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl you may have available.

=cut


lib/Device/BlinkyTape/SimulationPort.pm
package Device::BlinkyTape::SimulationPort;
use strict;
BEGIN {
    # AUTHORITY
    # VERSION
}
use Moose;
use Tk;

has 'baudrate' => (is => 'rw');
has 'databits' => (is => 'rw');
has 'parity' => (is => 'rw');
has 'stopbits' => (is => 'rw');

has 'simulate_window' => (is => 'rw');
has 'simulate_canvas' => (is => 'rw');
has 'port' => (is => 'rw');
has 'drawpixelpos' => (is => 'rw');
has 'pixels' => (is => 'rw');
has 'led_count' => (is => 'rw');

=for Pod::Coverage BUILD

=head1 Usage

This module replaces the Device::SerialPort device with a BlinkyTape simulator.
You can use it to develop for the BlinkyTape before actually getting the device.

=cut

sub BUILD {
    my $self = shift;
    my $pixelsize = 10;
    $self->simulate_window(MainWindow->new());
    $self->simulate_window()->title('BlinkyTape simulator');
    $self->simulate_window()->geometry($self->led_count()*$pixelsize . 'x' . ($pixelsize+200) . '-0+0'); # Canvas()->createRectangle(0,0,$self->led_count()*$pixelsize, $pixelsize);
    $self->simulate_window()->update();

    $self->simulate_canvas($self->simulate_window()->Canvas);
    $self->simulate_canvas()->pack(-expand=> 1, -fill => 'both');

    $self->simulate_canvas()->pack;
    $self->simulate_window()->update;

    my @pixel;
    for ($a=0; $a<=$self->led_count-1; $a++) {
        $pixel[$a] = $self->simulate_canvas()->createRectangle($a*$pixelsize,0,$a*$pixelsize+$pixelsize-1,$pixelsize, -fill => 'black', -width => 0);
    }
    $self->pixels(\@pixel);

    $self->drawpixelpos(0);
}

=head2 write

Write bytes to the simulated Port. These are interpreted as BlinkyTape colours and displayed on a Tk window.

=cut

sub write {
    my $self = shift;
    my $color = shift;
    my $r = unpack("x0 C1", $color);
    if ($r == 255) {
        $self->drawpixelpos(0);
        $self->simulate_window()->update();
        return 1;
    }
    my $g = unpack("x1 C1", $color);
    my $b = unpack("x2 C1", $color);
    my $rgbcolor = sprintf ("#%.2X%.2X%.2X", $r, $g, $b);
    my $pixel_rectangle_tkid = $self->pixels()->[$self->drawpixelpos];
    $self->simulate_canvas()->itemconfigure($pixel_rectangle_tkid, -fill => $rgbcolor);
    $self->drawpixelpos($self->drawpixelpos()+1);
    $self->simulate_canvas()->pack;
}

1;

=head1 AUTHOR

Oskari Okko Ojala E<lt>okko@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) Oskari Okko Ojala 2013

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl you may have available.

=cut


lib/Device/BlinkyTape.pm
package Device::BlinkyTape;
use strict;
BEGIN {
    # AUTHORITY
    # VERSION
}
use Moose;

if ($^O eq 'MSWin32') {
   require 'Win32::SerialPort';
   import Win32::SerialPort;
} else {
   require 'Device::SerialPort';
   import Device::SerialPort;
}

use Device::BlinkyTape::SimulationPort;
use Moose::Util::TypeConstraints;
use utf8;

=encoding utf-8
=cut

# ABSTRACT: Control a BlinkyTape led strip

=for Pod::Coverage BUILD

=head1 NAME

Device::BlinkyTape - Control a BlinkyTape led strip

=head1 SYNOPSIS

    use Device::BlinkyTape::WS2811; # BlinkyTape uses WS2811
    my $bb = Device::BlinkyTape::WS2811->new(dev => '/dev/tty.usbmodem');
    # Set all led pixels on full white
    $bb->all_on();
    # Set all led pixels off
    $bb->all_off();

    # Send a white pixel (RGB 255/255/255).
    # Pixels are sent one by one from left to right.
    $bb->send_pixel(255,255,255);
    # Send an RGB 5/5/5 pixel
    $bb->send_pixel(5,5,5);
    # Show all sent pixels and reset send_pixel to the first pixel.
    $bb->show();

=cut

=head1 SYNOPSIS, WINDOWS

    use Device::BlinkyTape::WS2811;
    use Win32::SerialPort;
    my $PortName = 'COM2';
    my $bb = Device::BlinkyTape::WS2811->new(Win32SerialPort => new Win32::SerialPort ($PortName));

    $bb->all_on();
    #...

=cut

# These Moose subtypes are defined for isa validation.
subtype 'GammaInt',
    as 'Int',
    where { $_ >= 0 and $_ <= 255 },
    message { "Gamma number ($_) must be between 0..255" }
;

subtype 'DeviceSerialPort',
    as 'Device::SerialPort'
;

subtype 'Win32SerialPort',
    as 'Win32::SerialPort'
;

subtype 'SimulationPort',
    as 'Device::BlinkyTape::SimulationPort'
;

=head2 dev

The device where your usb ledstrip is at. Defaults to /dev/tty.usbmodem.

=cut

has 'dev' => (is => 'rw', isa => 'Str', default => '/dev/tty.usbmodem');

=head2 port

Instead of giving the device you can create the instance of this module by directly
giving it a Device:SerialPort or a Win32::SerialPort object. By default the
Device::SerialPort object is created from the device given in the 'dev' parameter.

=cut

has 'port' => (is => 'rw', isa => 'DeviceSerialPort | SimulationPort | Win32SerialPort');

=head2 gamma

Specify the gamma correction. Defaults to [1,1,1]

=cut

has 'gamma' => (is => 'rw', isa => 'ArrayRef[GammaInt]', default => sub { [1,1,1] });

=head2 led_count

Specify the led count, counting from 1. The default is 60.

=cut

has 'led_count' => (is => 'rw', isa => 'Int', default => 60);

=head2 simulate

Specify if the module should simulate a BlinkyTape onscreen instead of using one in the usb port.
Defaults to 0 (false). If this is true then the port and dev parameters have no effect.

=cut

has 'simulate' => (is => 'rw', isa => 'Bool', default => 0);

=head2 sleeptime

Sending data too fast freezes the BlinkyTape. Sleeping for 30 microseconds between each byte
makes sure the atmega processor of the BlinkyTape can keep up with the incoming data.

=cut

has 'sleeptime' => (is => 'rw', default => 30);

sub BUILD {
    my $self = shift;
    # Initialize $self->port from $self->dev if one was not given in new
    if ($self->simulate) {
        $self->port(Device::BlinkyTape::SimulationPort->new(led_count => $self->led_count));
        warn 'simulation on.';
    }
    $self->port(Device::SerialPort->new($self->dev)) if (!$self->port);
    
    if ($self->port) {
        # Set default communication settings
        $self->port->baudrate(19200);
        $self->port->databits(8);
        $self->port->parity('none');
        $self->port->stopbits(1);
    }

    # $self->lookclear; # empty buffers
}

=head2 all_on

Turns all leds on.

=cut

sub all_on {
    my $self = shift;
    for (my $a=0; $a<=$self->led_count-1; $a++) {
        $self->send_pixel(255,255,255);
    }
    $self->show();
}

=head2 all_off

Turns all leds off.

=cut

sub all_off {
    my $self = shift;
    for (my $a=0; $a<=$self->led_count-1; $a++) {
        $self->send_pixel(0,0,0);
    }
    $self->show();
}

1;

=head2 send_pixel(r,g,b)

Send the RGB value for the next pixel. Values 0-254 are sent as is, value 255 is converted to 254.

=cut

=head2 show

Shows the sent pixels and resets the send_pixel to the first led pixel of the strip. This is done
by sending a single 255 value byte to the led strip.

=cut

=head1 ABOUT BLINKYTAPE

Blinkytape is a controllable led strip available at http://blinkiverse.com/blinkytape/

=cut

=head1 USING THE MODULE ON OS X WITHOUT OWNING A BLINKYTAPE

This module comes with a BlinkyTape simulator. Install X11 server to use the simulator:
    http://xquartz.macosforge.org/landing/

=cut

=head1 BUGS

The device is not yet available so the module has been implemented by inspecting partly
undocumented and unfinished code in other languages. Feel free to file any bug reports
in Github, patches welcome.

=cut

=head1 REFERENCE READING

Communicating with the Arduino in Perl http://playground.arduino.cc/interfacing/PERL

Perl communication to Arduino over serial USB http://www.windmeadow.com/node/38

=cut

=head1 AUTHOR

Oskari Okko Ojala E<lt>okko@cpan.orgE<gt>

Based on exampls/Blinkyboard.py in Blinkiverse's BlinkyTape repository at
https://github.com/blinkiverse/BlinkyTape/ 
by Max Henstell (mhenstell) and Marty McGuire (martymcguire).

=cut

=head1 COPYRIGHT AND LICENSE

Copyright (C) Oskari Okko Ojala 2013

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl you may have available.

=cut
